optimizely/event/factory.go

```diff
@@ -200,34 +200,34 @@ func createBatchEvent(userEvent UserEvent, visitor Visitor) Batch {
 }
 
 // get visitor attributes from user attributes
-func getEventAttributes(attributeKeyToIDMap map[string]string, attributes map[string]interface{}, botFiltering bool) []VisitorAttribute {
+func getEventAttributes(projectConfig optimizely.ProjectConfig, attributes map[string]interface{}) []VisitorAttribute {
 	var eventAttributes = []VisitorAttribute{}
 
 	for key, value := range attributes {
 		if value == nil {
 			continue
 		}
-		attribute := VisitorAttribute{}
-		id := attributeKeyToIDMap[key]
-		if id != "" {
-			attribute.EntityID = id
+		visitorAttribute := VisitorAttribute{}
+		attribute, _ := projectConfig.GetAttributeByKey(key)
+		if attribute.ID != "" {
+			visitorAttribute.EntityID = attribute.ID
 		} else if strings.HasPrefix(key, specialPrefix) {
-			attribute.EntityID = key
+			visitorAttribute.EntityID = key
 		} else {
 			continue
```

Comment:
If we reach this point does it mean the user passed in an invalid attribute? Should we log? IIRC in other SDKs we validate at a higher level, so would never expect this to happen when building an event.

Reply:
Good point. I'll see if we can take care of it before this. Otherwise we'll log here

Reply:
The filtering actually happens when building the event: https://github.com/optimizely/javascript-sdk/blob/master/packages/optimizely-sdk/lib/core/event_builder/index.js#L64 - which makes sense because we want to preserve the attributes when we send it back via the notification listener (for activate or track)